File: mobilefoundation/ios/Classes/WLAuthorizationManagerMethodHandler.swift
import Flutter
import IBMMobileFirstPlatformFoundation
import UIKit

/// This is a MethodCallManager class for WLAuthorizationManager method calls
public class WLAuthorizationManagerMethodHandler {
    static let shared = WLAuthorizationManagerMethodHandler()
    
    private init()
    {
        // Set up API instance
    }
    
    func handle(call: FlutterMethodCall, result: @escaping FlutterResult) {
        do {
            let isValid: Bool = try MFBridgeValidator.sharedInstance().validateMethodCall(call)
            if !isValid {
                result(FlutterError(code: MFConstants.ERR_INCORRECT_ARGS_CODE, message: MFConstants.ERR_INCORRECT_ARGS_MESSAGE, details: nil))
                return
            }
        } catch {
            print("Error \(error)")
            result(FlutterError(code: MFConstants.ERR_METHOD_NAME_CODE, message: error.localizedDescription, details: nil))
            return
        }
        let args = call.arguments as? [String: Any]
        switch call.method {
        case Methods.OBTAINACCESSTOKEN:
            let scope = args?[Arguments.SCOPE] as? String
            WLAuthorizationManager.sharedInstance()?.obtainAccessToken(forScope: scope, withCompletionHandler: { (token, error) -> Void in
                if error != nil {
                    let mfFailResponse = NSMutableDictionary()
                    mfFailResponse.setValue(String ((error! as NSError).code) , forKey: Arguments.ERROR_CODE)
                    mfFailResponse.setValue((error! as NSError).localizedDescription, forKey: Arguments.ERROR_MSG)
                    result(mfFailResponse)
                } else {
                    // Token is of AccessToken class -> converting it to a dictionary so that it can accessed as a MAP/Object in the dart layer.
                    let dictionary = NSMutableDictionary()
                    dictionary.setValue(token?.value, forKey: Arguments.VALUE)
                    dictionary.setValue(token?.asAuthorizationRequestHeaderField, forKey: Arguments.AS_AUTHORIZATION_REQUEST_HEADER)
                    dictionary.setValue(token?.asFormEncodedBodyParameter, forKey: Arguments.AS_FORMENCODED_BODY_PARAMETER)
                    dictionary.setValue(token?.scope, forKey: Arguments.SCOPE)
                    result(dictionary)
                }
            })
            break
        case Methods.CLEARACCESSTOKEN:
            let accessToken = args?[Arguments.ACCESSTOKEN] as? NSMutableDictionary
            let tokenToRemove = AccessToken(token: Arguments.ACCESSTOKEN, withExpiration: 0, forScope: accessToken?[Arguments.SCOPE] as? String)
            WLAuthorizationManager.sharedInstance()?.clear(tokenToRemove)
            result(nil)
            break
        case Methods.LOGIN:
            let securityCheck = args?[Arguments.SECURITYCHECK] as? String
            let credentials = args?[Arguments.CREDENTIALS] as? [AnyHashable: Any]
            WLAuthorizationManager.sharedInstance()?.login(securityCheck, withCredentials: credentials, withCompletionHandler: { (error) -> Void in
                if error != nil {
                    let mfFailResponse = NSMutableDictionary()
                    mfFailResponse.setValue(String ((error! as NSError).code) , forKey: Arguments.ERROR_CODE)
                    mfFailResponse.setValue((error! as NSError).localizedDescription, forKey: Arguments.ERROR_MSG)
                    result(mfFailResponse)
                } else {
                    result(nil)
                }
            })
            break
        case Methods.LOGOUT:
            let securityCheck = args?[Arguments.SECURITYCHECK] as? String
            WLAuthorizationManager.sharedInstance()?.logout(securityCheck, withCompletionHandler: { (error) -> Void in
                if error != nil {
                    let mfFailResponse = NSMutableDictionary()
                    mfFailResponse.setValue(String ((error! as NSError).code) , forKey: Arguments.ERROR_CODE)
                    mfFailResponse.setValue((error! as NSError).localizedDescription, forKey: Arguments.ERROR_MSG)
                    result(mfFailResponse)
                } else {
                    result(nil)
                }
            })
            break
        case Methods.SETLOGINTIMEOUT:
            let timeOut = args?[Arguments.TIMEOUT] as? NSNumber
            WLAuthorizationManager.sharedInstance()?.setLoginTimeout(timeOut)
            result(nil)
            break
        default:
            result(FlutterError(code: MFConstants.ERR_METHOD_NAME_CODE, message: MFConstants.ERR_METHOD_NAME_MESSAGE, details: nil))
        }
    }
    
    private enum Methods{
        public static var OBTAINACCESSTOKEN: String = "obtainAccessToken"
        public static var CLEARACCESSTOKEN: String = "clearAccessToken"
        public static var LOGIN: String = "login"
        public static var LOGOUT: String = "logout"
        public static var SETLOGINTIMEOUT: String = "setLoginTimeOut"
    }
    
    private enum Arguments{
        public static var STATUS: String = "status"
        public static var HEADERS: String = "headers"
        public static var ERROR_CODE: String = "errorCode"
        public static var ERROR_MSG: String = "errorMsg"
        public static var SCOPE: String = "scope"
        public static var ACCESSTOKEN: String = "accessToken"
        public static var VALUE: String = "value"
        public static var AS_AUTHORIZATION_REQUEST_HEADER: String = "asAuthorizationRequestHeader"
        public static var AS_FORMENCODED_BODY_PARAMETER: String = "asFormEncodedBodyParameter"
        public static var SECURITYCHECK: String = "securityCheck"
        public static var CREDENTIALS: String = "credentials"
        public static var TIMEOUT: String = "timeOut"
    }
}

File: mobilefoundation/android/src/main/java/com/ibm/mobile/mobilefoundation/WLAuthorizationManagerMethodHandler.java
package com.ibm.mobile.mobilefoundation;

import android.app.Activity;

import androidx.annotation.NonNull;

import com.worklight.wlclient.api.WLAccessTokenListener;
import com.worklight.wlclient.api.WLAuthorizationManager;
import com.worklight.wlclient.api.WLFailResponse;
import com.worklight.wlclient.api.WLLoginResponseListener;
import com.worklight.wlclient.api.WLLogoutResponseListener;
import com.worklight.wlclient.auth.AccessToken;
import com.worklight.wlclient.auth.WLAuthorizationManagerInternal;

import org.json.JSONObject;

import java.net.MalformedURLException;
import java.net.URL;
import java.util.HashMap;
import java.util.Map;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;

public class WLAuthorizationManagerMethodHandler implements MethodChannel.MethodCallHandler {

    private static WLAuthorizationManagerMethodHandler instance;
    private final WLAuthorizationManager wlAuthorizationManager = WLAuthorizationManager.getInstance();
    private Activity activity;

    private WLAuthorizationManagerMethodHandler() {
    }

    public static WLAuthorizationManagerMethodHandler getInstance() {
        if (instance == null) {
            instance = new WLAuthorizationManagerMethodHandler();
        }
        return instance;
    }


    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull final MethodChannel.Result result) {
        try {
            boolean isValid = MFBridgeValidator.getInstance().validateMethodCall(call);
            if (!isValid) {
                result.error(MFConstants.ERR_INCORRECT_ARGS_CODE, MFConstants.ERR_INCORRECT_ARGS_MESSAGE, call.method);
                return;
            }
        } catch (MFBridgeValidator.MethodNotImplementedException ex) {
            result.error(MFConstants.ERR_METHOD_NAME_CODE, MFConstants.ERR_METHOD_NAME_MESSAGE, call.method);
            return;
        }

        if (activity == null) {
            activity = MobilefoundationPlugin.getActivity();
        }

        switch (call.method) {
            case Methods.OBTAIN_ACCESS_TOKEN:
                wlAuthorizationManager.obtainAccessToken((String) call.argument(Arguments.SCOPE), new WLAccessTokenListener() {
                    @Override
                    public void onSuccess(final AccessToken accessToken) {
                        activity.runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                Map<String, String> token = new HashMap<String, String>();
                                token.put(MFConstants.VALUE, accessToken.getValue());
                                token.put(MFConstants.SCOPE, accessToken.getScope());
                                token.put(MFConstants.AS_AUTHORIZATION_REQUEST_HEADER, accessToken.getAsAuthorizationRequestHeader());
                                token.put(MFConstants.AS_FORMENCODED_BODY_PARAMETER, accessToken.getAsFormEncodedBodyParameter());
                                result.success(token);
                            }
                        });
                    }

                    @Override
                    public void onFailure(final WLFailResponse wlFailResponse) {
                        activity.runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                Map<String, Object> mfFailResponse = new HashMap();
                                mfFailResponse.put(Arguments.STATUS, wlFailResponse.getStatus());
                                mfFailResponse.put(Arguments.HEADERS, wlFailResponse.getHeaders());
                                mfFailResponse.put(Arguments.ERROR_CODE, wlFailResponse.getErrorStatusCode());
                                mfFailResponse.put(Arguments.ERROR_MSG, wlFailResponse.getErrorMsg());
                                result.success(mfFailResponse);
                            }
                        });
                    }
                });
                break;
            case Methods.WLAUTHORIZATIONMANAGER_CLEARACCESSTOKEN :
                String scope = ((HashMap<String, String>)call.argument(Arguments.ACCESSTOKEN)).get(MFConstants.SCOPE);
                WLAuthorizationManagerInternal.getInstance().removeTokenByScope(scope);
                result.success(null);
                break;

            case Methods.WLAUTHORIZATIONMANAGER_LOGIN:
                wlAuthorizationManager.login((String) call.argument(Arguments.SECURITYCHECK), new JSONObject(((HashMap<String, String>) call.argument(Arguments.CREDENTIALS))), new WLLoginResponseListener() {
                    @Override
                    public void onSuccess() {
                        activity.runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                result.success(null);
                            }
                        });
                    }

                    @Override
                    public void onFailure(final WLFailResponse wlFailResponse) {
                        activity.runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                Map<String, Object> mfFailResponse = new HashMap();
                                mfFailResponse.put(Arguments.STATUS, wlFailResponse.getStatus());
                                mfFailResponse.put(Arguments.HEADERS, wlFailResponse.getHeaders());
                                mfFailResponse.put(Arguments.ERROR_CODE, wlFailResponse.getErrorStatusCode());
                                mfFailResponse.put(Arguments.ERROR_MSG, wlFailResponse.getErrorMsg());
                                result.success(mfFailResponse);
                            }
                        });
                    }
                });
                break;

            case Methods.WLAUTHORIZATIONMANAGER_LOGOUT:
                wlAuthorizationManager.logout((String) call.argument(Arguments.SECURITYCHECK), new WLLogoutResponseListener() {
                    @Override
                    public void onSuccess() {
                        activity.runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                result.success(null);
                            }
                        });
                    }

                    @Override
                    public void onFailure(final WLFailResponse wlFailResponse) {
                        activity.runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                Map<String, Object> mfFailResponse = new HashMap();
                                mfFailResponse.put(Arguments.STATUS, wlFailResponse.getStatus());
                                mfFailResponse.put(Arguments.HEADERS, wlFailResponse.getHeaders());
                                mfFailResponse.put(Arguments.ERROR_CODE, wlFailResponse.getErrorStatusCode());
                                mfFailResponse.put(Arguments.ERROR_MSG, wlFailResponse.getErrorMsg());
                                result.success(mfFailResponse);
                            }
                        });
                    }
                });
                break;

            case Methods.WLAUTHORIZATIONMANAGER_SETLOGINTIMEOUT :
                wlAuthorizationManager.setLoginTimeout((int)call.argument(Arguments.TIMEOUT));
                result.success(null);
                break;

            default:
                result.error(MFConstants.ERR_METHOD_NAME_CODE, MFConstants.ERR_METHOD_NAME_MESSAGE, null);
        }
    }

    private static class Methods {
        public static final String OBTAIN_ACCESS_TOKEN = "obtainAccessToken";
        public static final String WLAUTHORIZATIONMANAGER_CLEARACCESSTOKEN = "clearAccessToken";
        public static final String WLAUTHORIZATIONMANAGER_LOGIN = "login";
        public static final String WLAUTHORIZATIONMANAGER_LOGOUT = "logout";
        public static final String WLAUTHORIZATIONMANAGER_SETLOGINTIMEOUT  = "setLoginTimeOut";
    }

    private static class Arguments {
        public static final String SCOPE = "scope";
        public static final String ACCESSTOKEN = "accessToken";
        public static final String SECURITYCHECK = "securityCheck";
        public static final String CREDENTIALS = "credentials";
        public static final String TIMEOUT = "timeOut";
        public static final String ERROR_CODE = "errorCode";
        public static final String ERROR_MSG  = "errorMsg";
        public static final String HEADERS = "headers";
        public static final String STATUS = "status";
    }
}
